### ISSUE TITLE
Reporting Latency

### ISSUE THREAD

**Issue by ekaynar** [awslabs/aws-crt-s3-benchmarks#64]
Do you plan to add a feature for the benchmark to report the average request latency?

**Comment by ekaynar** [comment]
I am particularly interested in end-to-end latency (HTTP-request-send-begin-to-response-recv-complete), particularly for S3 Express. 

Is there an example you can point me to showing how I can hook telemetry_callback to S3 request in crt-c runner?

**Comment by graebm** [comment]
This repo mostly benchmarks "transfer managers" which will split up a download or upload into many parallel HTTP requests. So if you're uploading a large file, the transfer manager will do [multipart upload](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html#mpu-process), sending N+2 HTTP requests: [StartMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html), then N [UploadPart](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html), and finally [CompleteMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html). You're going to see very different latency from each type of request (CompleteMultipartUpload is slowest, despite it transferring the least amount of data)

Anyway, what I'm getting at is: if you simply want to measure latency for regular old [GetObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html) or [PutObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html), where it's reliably doing 1 HTTP request per object, you'd be better off starting with a code sample using something well-documented like the [AWS SDK for Java v2](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3.html). See [Use SDK metrics from the AWS SDK for Java](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/metrics.html)

Or are you really specifically interested in S3 transfer managers?

**Comment by graebm** [comment]
If you are still interested in hooking up the telemetry_callback in s3-benchrunner-c, here's a branch where someone was messing around, writing some specific stats to a `.csv` file: https://github.com/awslabs/aws-crt-s3-benchmarks/compare/main...checksum-test#diff-79e3bfecfecf5f534af2219854efb9ddd44b05d3f9f352d0ef9d758b1c976e2b

**Comment by ekaynar** [comment]
Thank you for sharing the details. Very useful. I modified the code to use classic SDK (not CRT) to capture the latency numbers. 

I observed that when I use the SDK based on CRT, it is still faster than the classic SDK, even for small object sizes such (<1MB - which I assume CRT does not use multipart download/upload).

Anyway, I will play more with the CRT libraries. Thanks again for the help and clarification of the technical details.
